### PR TITLE
Ensure the `*_fields` attributes of `__init__`/`postinit` are cohesive

### PR DESCRIPTION
**Pull request by thejcannon**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description
It was discovered (and the fix is needed by) #1194 that there isn't strict cohesion between the `*_fields` attributes and the parameters in `__init__` and `postinit`, so I've added tests which enforces this.

Through testing a handful of violators have been found, so a deprecation warning decorator has been added, and parameters have been updated accordingly.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
|    | :scroll: Docs          |

### CODE REVIEW

**Comment by thejcannon** [comment]
It says in https://github.com/PyCQA/astroid/blob/3f98fa0be2279ff422a0e157705f4e3ea1e4d48f/doc/extending.rst that 

> When instantiating a node, the non-AST parameters are usually passed via the constructor, while the AST parameters are provided via the postinit() method. The only exception is that the parent is also passed via the constructor. Instantiating a new node might look as in:

So in addition to having these follow a convention for developer's sake conventions also helps allow code to make assumptions.

So I wouldn't say it's "just for the point of doing it". It's to address some tech debt / code uncleanliness.

File: astroid/nodes/node_classes.py [hunk 1]
```diff
@@ -1328,7 +1332,7 @@ def postinit(
         self.target = target
         self.annotation = annotation
         self.value = value
-        self.simple = simple
+        self.simple = simple or self.simple
```
[the remark this answers is not in the document]
Comment (thejcannon) [reply astroid/nodes/node_classes.py:1335]: Yeah this was a forehead-slapper.

**Comment by cdce8p** [comment]
> It says in https://github.com/PyCQA/astroid/blob/3f98fa0be2279ff422a0e157705f4e3ea1e4d48f/doc/extending.rst that
> 
> > When instantiating a node, the non-AST parameters are usually passed via the constructor, while the AST parameters are provided via the postinit() method. The only exception is that the parent is also passed via the constructor. Instantiating a new node might look as in:
> 
> So in addition to having these follow a convention for developer's sake conventions also helps allow code to make assumptions.
> 
> So I wouldn't say it's "just for the point of doing it". It's to address some tech debt / code uncleanliness.

We might need to update the documentation then 😉
I still don't think we should change it. Even if it addresses some tech debt, there are also good arguments to keep it the way it's now. The most important is the risk of regressions: Just from the current changes, we would have introduced three new regressions that could each crash astroid and by extension pylint. It's just not worth it!

**Comment by Pierre-Sassoulas** [comment]
We've done some pretty heavy refactor and breaking change in 3.0 for consistency's sake, Marc, what is different here ? It think having a clear rule like:
- non-AST parameters are passed via the constructor # notice I removed "usually"
- AST parameters are provided via the postinit()
- except for "parents" which are passed to constructor

Would make astroid development easier and reduce mistakes then the need to retro-engineer classes to see how they were implemented if they do not follow the general "rule". Is a breaking change in 3.0 really that much more impacting than what we already did ?

**Comment by cdce8p** [comment]
> We've done some pretty heavy refactor and breaking change in 3.0 for consistency's sake, Marc, what is different here ?

The breaking changes so far have been because we moved code around. So changing the import is usually enough. Then there are the cases which shouldn't happen anyway, like passing `None` as argument in cases where it should never be `None`. Those are bugs that we previously handled gracefully but to improve typing, we deprecate it now.

This change in contrast will break code just because it "might" improve consistency. IMO this is dangerous and not worth it. As I've pointed out already, just from a few lines, there would have been multiple regressions. True, we can fix them but who is to say those don't introduce new bugs in plugins. Usually I think it's better to limit breaking changes to an absolute minimum and only do them when there is a clear benefit.

To highlight something different. Although it might fix inconsistencies, it also introduces new ones. All node types have `lineno`, `col_offset`, and `parent` as their last arguments for `__init__`. This PR creates an exceptions just for a few. Should we mention that in the docs, then?

A last point, the `ast` changes with every Python release. These inconsistencies are a direct result of that. We can easily add new arguments to `postinit` whereas with `__init__` we would create breaking changes or change the order, i.e. `parent` isn't the last anymore. One of the major things `astroid` does so well, is hide the `ast` behind a somewhat stable interface `pylint` and plugins can work with. As for the need to "retro-engineer" classes, I would be surprised if anyone is able to write `plugins` without at least looking at the documentation, somethings even that isn't enough. This PR won't change that.

**Comment by Pierre-Sassoulas** [comment]
Ok, I'm convinced, it seemed really nice on paper @thejcannon thank you for the proposal. But the inconsistencies coming from the AST that will keep on coming and the bigger changes required for lib that are not us (pylint) are pretty convincing reasons to not fix this.

**Comment by thejcannon** [comment]
Bummer. Cheers anyways.

**Pull request closed without merging**